targets/rpc.cpp

```diff
@@ -0,0 +1,450 @@
+#include <memory>
+#include <optional>
+#include <stdexcept>
+#include <vector>
+enum class ChainType;
+
+using util::Join;
+using util::ToString;
+
+extern "C" int go_btcd_rpc(char* cargs, int len);
+
+namespace {
+struct RPCFuzzTestingSetup : public TestingSetup {
+
+    RPCFuzzTestingSetup(const ChainType chain_type, TestOpts opts) : TestingSetup{chain_type, opts}
+    {
+    }
+
+    int CallRPC(const JSONRPCRequest& request)
+    {
+        auto ret = JSONRPCExec(request, true);
+
+        if (ret.write().find("unknown") != std::string::npos)
+            return 1;
+
+        if (ret["error"].isNull())
+            return 0;
+
+        if (ret["error"]["code"].getInt<int64_t>() == -1)
+            return -1;
+
+        return 1;
+    }
+
+    int CallRPCBtcd(const JSONRPCRequest& request)
+    {
+        const auto& str = JSONRPCRequestObj(request.strMethod, request.params, 0).write();
+        return go_btcd_rpc((char*)str.c_str(), str.size());
+    }
+
+    JSONRPCRequest GenerateRequest(const std::string& rpc_method, const std::vector<std::string>& arguments)
+    {
+        JSONRPCRequest request;
+        request.context = &m_node;
+        request.strMethod = rpc_method;
+        request.params = RPCConvertValues(rpc_method, arguments);
+
+        return request;
+    }
+
+    std::vector<std::string> GetRPCCommands() const
+    {
+        return tableRPC.listCommands();
+    }
+};
+
+RPCFuzzTestingSetup* rpc_testing_setup = nullptr;
+std::string g_limit_to_rpc_command;
+std::unique_ptr<RPCFuzzTestingSetup> setup;
+
+// RPC commands which are not appropriate for fuzzing: such as RPC commands
+// reading or writing to a filename passed as an RPC parameter, RPC commands
+// resulting in network activity, etc.
+const std::vector<std::string> RPC_COMMANDS_NOT_SAFE_FOR_FUZZING{
+    "addconnection",  // avoid DNS lookups
+    "addnode",        // avoid DNS lookups
+    "addpeeraddress", // avoid DNS lookups
+    "dumptxoutset",   // avoid writing to disk
+    "dumpwallet", // avoid writing to disk
+    "enumeratesigners",
+    "echoipc",              // avoid assertion failure (Assertion `"EnsureAnyNodeContext(request.context).init" && check' failed.)
+    "generatetoaddress",    // avoid prohibitively slow execution (when `num_blocks` is large)
+    "generatetodescriptor", // avoid prohibitively slow execution (when `nblocks` is large)
+    "gettxoutproof",        // avoid prohibitively slow execution
+    "importmempool", // avoid reading from disk
+    "importwallet", // avoid reading from disk
+    "loadtxoutset",   // avoid reading from disk
+    "loadwallet",   // avoid reading from disk
+    "savemempool",           // disabled as a precautionary measure: may take a file path argument in the future
+    "setban",                // avoid DNS lookups
+    "stop",                  // avoid shutdown state
+};
+
+// RPC commands which are safe for fuzzing.
+const std::vector<std::string> RPC_COMMANDS_SAFE_FOR_FUZZING{
```

Comment:
Did you check if btcd has all these RPCs?

Reply:
Hi, btcd is definitely missing some of these RPCs, however that is not a problem, since btcd in such cases returns a very specific error code which we check before the asserts